iree/hal/metal/metal_command_buffer.mm

```diff
@@ -100,7 +215,29 @@
                                       Buffer* target_buffer, device_size_t target_offset,
                                       device_size_t length) {
   IREE_TRACE_SCOPE0("MetalCommandBuffer::CopyBuffer");
-  return UnimplementedErrorBuilder(IREE_LOC) << "MetalCommandBuffer::CopyBuffer";
+
+  IREE_ASSIGN_OR_RETURN(auto* source_device_buffer, CastBuffer(source_buffer));
+  IREE_ASSIGN_OR_RETURN(auto* target_device_buffer, CastBuffer(target_buffer));
+
+  source_offset += source_buffer->byte_offset();
+  target_offset += target_buffer->byte_offset();
+
+#ifdef IREE_PLATFORM_MACOS
+  // Per the spec for copyFromBuffer:sourceOffset:toBuffer:destinationOffset:size, the source/target
```

Comment:
this kind of logic scares me - we should uniformly fail in all cases as much as possible, otherwise there's a giant pit of sadness waiting when trying to go from running on linux to running on ios (and debugging on mac wouldn't help). I'd rather require that all offsets/lengths are 4-byte aligned in all backends then special case this.

Reply:
(if we need to polyfill non-byte fills, then having our own dispatches here for non-word-aligned copies seems ok - and could be TODO'd to teh same issue)

Reply:
Sounds good to me. I was trying to gain the best from each target. But yeah we control HAL entirely; padding a few bytes gives us the win for uniformity and that's good. 

Removed the `#ifdef`.